cli/src/main.rs

```diff
@@ -165,8 +169,14 @@ fn main() -> Result<()> {
             base_address,
             skip_bytes,
             path,
-        } => download_program_fast(common, format.into(base_address, skip_bytes), &path),
-        Cli::Erase { common } => erase(&common),
+            erase,
+        } => {
+            if erase {
+                erase_device(&common)?;
+            }
+            download_program_fast(common, format.into(base_address, skip_bytes), &path)
```

Comment:
Currently this is doing a chip erase, then the flashing is doing page erases again.

If you set [chip_erase](https://github.com/probe-rs/probe-rs/blob/3d492bab765aa8cda71e3ecdb39377f1b84df15e/probe-rs/src/flashing/download.rs#L110) to `true` in the flashing options, it'll do one chip erase then write the data without page erases, which is much faster.

Reply:
I thought I remembered that but download uses run_flash_download though which doesnt use the downloadoptions I think

Reply:
Oh I see it

Reply:
Do you like sticking it in the flash common options then or just another arg

https://github.com/probe-rs/probe-rs/blob/3d492bab765aa8cda71e3ecdb39377f1b84df15e/cli/src/main.rs#L248

Reply:
This is stuck forever on the progress bar after programming, querying somthing from the chip. Both a daplink and jlink.

In chat dirbaio found it doesnt appear to be the probe, but rather the progress bars


> here https://github.com/probe-rs/probe-rs/blob/master/probe-rs-cli-util/src/flash.rs#L141
so I guess some progress bar is not "finishing"
fun :S
well at least it's not an actual probe issue

```
(gdb) i thread
  Id   Target Id                                         Frame 
* 1    Thread 0x7ffff7c09800 (LWP 155588) "probe-rs-cli" 0x00007ffff7f3a8ca in __futex_abstimed_wait_common64 () from /usr/lib/libpthread.so.0
  4    Thread 0x7ffff7c08640 (LWP 155594) "probe-rs-cli" 0x00007ffff7cffb2f in poll () from /usr/lib/libc.so.6
  5    Thread 0x7ffff735d640 (LWP 155595) "probe-rs-cli" 0x00007ffff7d0518d in syscall () from /usr/lib/libc.so.6
(gdb) t 1
[Switching to thread 1 (Thread 0x7ffff7c09800 (LWP 155588))]
#0  0x00007ffff7f3a8ca in __futex_abstimed_wait_common64 () from /usr/lib/libpthread.so.0
(gdb) bt
#0  0x00007ffff7f3a8ca in __futex_abstimed_wait_common64 () from /usr/lib/libpthread.so.0
#1  0x00007ffff7f2f743 in __pthread_clockjoin_ex () from /usr/lib/libpthread.so.0
#2  0x00005555561ab67d in std::sys::unix::thread::Thread::join () at library/std/src/sys/unix/thread.rs:242
#3  0x000055555590fbd5 in std::thread::JoinInner<()>::join<()> (self=0x7fffffffbc40)
    at /rustc/c1026539bd22e9d070988deaa47b1360cbc76436/library/std/src/thread/mod.rs:1270
#4  0x000055555590fc51 in std::thread::JoinHandle<()>::join<()> (self=...) at /rustc/c1026539bd22e9d070988deaa47b1360cbc76436/library/std/src/thread/mod.rs:1402
#5  0x000055555590f5df in probe_rs_cli_util::flash::run_flash_download (session=0x7fffffffc4c0, path=0x555556da9480, opt=0x7fffffffd1b0, loader=..., 
    do_chip_erase=true) at probe-rs-cli-util/src/flash.rs:158
#6  0x0000555555870210 in probe_rs_cli::download_program_fast (common=..., format=..., path=..., do_chip_erase=true) at cli/src/main.rs:260
#7  0x000055555586ecc9 in probe_rs_cli::main () at cli/src/main.rs:173
(gdb) t 4
[Switching to thread 4 (Thread 0x7ffff7c08640 (LWP 155594))]
#0  0x00007ffff7cffb2f in poll () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007ffff7cffb2f in poll () from /usr/lib/libc.so.6
#1  0x00005555560e3b31 in linux_udev_event_thread_main (arg=0x0) at libusb/libusb/os/linux_udev.c:183
#2  0x00007ffff7f2e259 in start_thread () from /usr/lib/libpthread.so.0
#3  0x00007ffff7d0a5e3 in clone () from /usr/lib/libc.so.6
(gdb) t 5
[Switching to thread 5 (Thread 0x7ffff735d640 (LWP 155595))]
#0  0x00007ffff7d0518d in syscall () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007ffff7d0518d in syscall () from /usr/lib/libc.so.6
#1  0x00005555561a508a in std::sys::unix::futex::futex_wait () at library/std/src/sys/unix/futex.rs:25
#2  std::sys_common::thread_parker::futex::Parker::park () at library/std/src/sys_common/thread_parker/futex.rs:50
#3  std::thread::park () at library/std/src/thread/mod.rs:905
#4  std::sync::mpsc::blocking::WaitToken::wait () at library/std/src/sync/mpsc/blocking.rs:68
#5  0x0000555555941a29 in std::sync::mpsc::shared::Packet<(usize, indicatif::state::ProgressDrawState)>::recv<(usize, indicatif::state::ProgressDrawState)> (
    self=0x5555573b2970, deadline=...) at /rustc/c1026539bd22e9d070988deaa47b1360cbc76436/library/std/src/sync/mpsc/shared.rs:234
#6  0x000055555593433e in std::sync::mpsc::Receiver<(usize, indicatif::state::ProgressDrawState)>::recv<(usize, indicatif::state::ProgressDrawState)> (
    self=0x7ffff735c948) at /rustc/c1026539bd22e9d070988deaa47b1360cbc76436/library/std/src/sync/mpsc/mod.rs:1170
#7  0x0000555555938433 in indicatif::progress_bar::MultiProgress::join_impl (self=0x7ffff735c930, clear=false)
    at /home/dirbaio/.cargo/registry/src/github.com-1ecc6299db9ec823/indicatif-0.16.2/src/progress_bar.rs:682
#8  0x0000555555938040 in indicatif::progress_bar::MultiProgress::join (self=0x7ffff735c930)
    at /home/dirbaio/.cargo/registry/src/github.com-1ecc6299db9ec823/indicatif-0.16.2/src/progress_bar.rs:655
#9  0x0000555555909072 in probe_rs_cli_util::flash::run_flash_download::{closure#1} () at probe-rs-cli-util/src/flash.rs:143
#10 0x00005555559159fc in std::sys_common::backtrace::__rust_begin_short_backtrace<probe_rs_cli_util::flash::run_flash_download::{closure#1}, ()> (f=...)
    at /rustc/c1026539bd22e9d070988deaa47b1360cbc76436/library/std/src/sys_common/backtrace.rs:123
#11 0x0000555555910981 in std::thread::{impl#0}::spawn_unchecked::{closure#1}::{closure#0}<probe_rs_cli_util::flash::run_flash_download::{closure#1}, ()> ()
    at /rustc/c1026539bd22e9d070988deaa47b1360cbc76436/library/std/src/thread/mod.rs:483
#12 0x0000555555926d01 in core::panic::unwind_safe::{impl#23}::call_once<(), std::thread::{impl#0}::spawn_unchecked::{closure#1}::{closure#0}> (self=..., _args=())
    at /rustc/c1026539bd22e9d070988deaa47b1360cbc76436/library/core/src/panic/unwind_safe.rs:271
#13 0x0000555555926653 in std::panicking::try::do_call<core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked::{closure#1}::{closure#0}>, ()> (data=0x7ffff735cad8) at /rustc/c1026539bd22e9d070988deaa47b1360cbc76436/library/std/src/panicking.rs:403
#14 0x000055555592671b in __rust_try ()
#15 0x0000555555926591 in std::panicking::try<(), core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked::{closure#1}::{closure#0}>> (f=...)
    at /rustc/c1026539bd22e9d070988deaa47b1360cbc76436/library/std/src/panicking.rs:367
#16 0x0000555555922e01 in std::panic::catch_unwind<core::panic::unwind_safe::AssertUnwindSafe<std::thread::{impl#0}::spawn_unchecked::{closure#1}::{closure#0}>, ()> (
    f=...) at /rustc/c1026539bd22e9d070988deaa47b1360cbc76436/library/std/src/panic.rs:133
#17 0x000055555591079d in std::thread::{impl#0}::spawn_unchecked::{closure#1}<probe_rs_cli_util::flash::run_flash_download::{closure#1}, ()> ()
    at /rustc/c1026539bd22e9d070988deaa47b1360cbc76436/library/std/src/thread/mod.rs:482
#18 0x0000555555918b8f in core::ops::function::FnOnce::call_once<std::thread::{impl#0}::spawn_unchecked::{closure#1}, ()> ()
    at /rustc/c1026539bd22e9d070988deaa47b1360cbc76436/library/core/src/ops/function.rs:227
#19 0x00005555561ab5a3 in alloc::boxed::{impl#44}::call_once<(), dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global> ()
    at /rustc/c1026539bd22e9d070988deaa47b1360cbc76436/library/alloc/src/boxed.rs:1691
#20 alloc::boxed::{impl#44}::call_once<(), alloc::boxed::Box<dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global>, alloc::alloc::Global> ()
    at /rustc/c1026539bd22e9d070988deaa47b1360cbc76436/library/alloc/src/boxed.rs:1691
#21 std::sys::unix::thread::{impl#2}::new::thread_start () at library/std/src/sys/unix/thread.rs:106
#22 0x00007ffff7f2e259 in start_thread () from /usr/lib/libpthread.so.0
#23 0x00007ffff7d0a5e3 in clone () from /usr/lib/libc.so.6
```

Reply:
Indeed, I added `--disable-progressbars` to this commit, and that correctly exits

Reply:
Found it bug in `FlashLoader` not emitting `ProgressEvent::FinishedErasing` when using `do_chip_erase`.